### DISCUSSION TITLE
Why is error `FAIL. Reason: Contract 0xb4c79daB8f259C7Aee6E5b2Aa729821864227e84 does not exists` appearing in GitHub Actions?

### DISCUSSION THREAD

**Discussion by minaminao** [foundry-rs/foundry#2534] [Help] [answered]
The error has occurred [here](https://github.com/minaminao/ctf-blockchain/runs/7608015335?check_suite_focus=true).

Running `forge test` locally does not result in this error.

According to https://book.getfoundry.sh/forge/writing-tests?highlight=0xb4c79daB8f259C7Aee6E5b2Aa729821864227e84 , this address is where the test contract is deployed.

**Comment by minaminao** [comment]
It appears that this error occurs when calling the public library; the error does not occur with the internal library.

**Reply by minaminao** [reply]
In this case, the internal library is sufficient, so I changed it to the internal library and the error stopped, but why does the public library cause the error in GitHub Actions even though it is fine in the local environment?
https://github.com/minaminao/ctf-blockchain/commit/a073a6422db32bcb296c7e88bc20197b29964c3c

**Comment by onbjerg** [comment, marked as answer]
I'd assume your local installation is out of date vs CI. Some changes were made to how forking works recently, @mattsse has more insight on that

**Reply by minaminao** [reply]
Ah, I did `foundryup` and now I get the error locally too. Thanks. 
However, if someone could tell me why linked libraries give the error, that would be great.

Error:
```txt
[FAIL. Reason: Contract 0xb4c79daB8f259C7Aee6E5b2Aa729821864227e84 does not exists on active fork with id `1`
        But exists on non active forks: `[0]`]
```

**Reply by onbjerg** [reply]
I think this is due to how persistence is now handled in a multi-fork environment. I'm not sure if you use the fork-related cheatcodes, if not, this is likely a bug, otherwise read this section: https://book.getfoundry.sh/forge/fork-testing#separated-and-persistent-storage (and maybe the text before that)

**Reply by minaminao** [reply]
Probably a bug. The simplified code is: https://github.com/minaminao/foundry-linked-library-bug

Running `forge test` results in the error.

**Reply by minaminao** [reply]
btw, I tried to experiment with persistence, but they were not in [Vm.sol](https://github.com/foundry-rs/forge-std/blob/master/src/Vm.sol).

**Reply by onbjerg** [reply]
Please file issues on both of them in their respective repositories.

**Reply by minaminao** [reply]
done :)
#2549
https://github.com/foundry-rs/forge-std/issues/142
https://github.com/foundry-rs/forge-std/pull/143 (PR)
